plugin/runmetrics/producer.go

```diff
@@ -0,0 +1,282 @@
+package runmetrics
+
+import (
+	"errors"
+	"go.opencensus.io/metric"
+	"go.opencensus.io/metric/metricdata"
+	"go.opencensus.io/metric/metricproducer"
+	"runtime"
+)
+
+type (
+	// producer produces runtime metrics.
+	//
+	// Enable collection of runtime metrics with Enable().
+	producer struct {
+		options RunMetricOptions
+		reg     *metric.Registry
+
+		memStats *memStats
+		cpuStats *cpuStats
+	}
+
+	// RunMetricOptions allows to configure runtime metrics.
+	RunMetricOptions struct {
+		EnableCPU    bool   // EnableCPU whether CPU metrics shall be recorded
+		EnableMemory bool   // EnableMemory whether memory metrics shall be recorded
+		Prefix       string // Prefix is a custom prefix for metric names
+	}
+
+	memStats struct {
+		memStats runtime.MemStats
+
+		memAlloc   *metric.Int64GaugeEntry
+		memTotal   *metric.Int64GaugeEntry
+		memSys     *metric.Int64GaugeEntry
+		memLookups *metric.Int64GaugeEntry
+		memMalloc  *metric.Int64GaugeEntry
+		memFrees   *metric.Int64GaugeEntry
+
+		heapAlloc    *metric.Int64GaugeEntry
+		heapSys      *metric.Int64GaugeEntry
+		heapIdle     *metric.Int64GaugeEntry
+		heapInuse    *metric.Int64GaugeEntry
+		heapObjects  *metric.Int64GaugeEntry
+		heapReleased *metric.Int64GaugeEntry
+
+		stackInuse       *metric.Int64GaugeEntry
+		stackSys         *metric.Int64GaugeEntry
+		stackMSpanInuse  *metric.Int64GaugeEntry
+		stackMSpanSys    *metric.Int64GaugeEntry
+		stackMCacheInuse *metric.Int64GaugeEntry
+		stackMCacheSys   *metric.Int64GaugeEntry
+	}
+
+	cpuStats struct {
+		numGoroutines *metric.Int64GaugeEntry
+		numCgoCalls   *metric.Int64GaugeEntry
+	}
+)
+
+var _ metricproducer.Producer = (*producer)(nil)
+
+// Enable creates and enables a new runtime metrics producer.
+//
+// Supply RunMetricOptions to configure the behavior of the producer.
+// An error might be returned, if creating metrics gauges fails.
+//
+// Previous calls will be overwritten by subsequent ones.
+func Enable(options RunMetricOptions) error {
+	producer := &producer{options: options, reg: metric.NewRegistry()}
+	var err error
+
+	if options.EnableMemory {
+		producer.memStats, err = newMemStats(producer)
+		if err != nil {
+			return err
+		}
+	}
+
+	if options.EnableCPU {
+		producer.cpuStats, err = newCPUStats(producer)
+		if err != nil {
+			return err
+		}
+	}
+
+	Disable() // ensure no duplicate registrations
+	metricproducer.GlobalManager().AddProducer(producer)
+	return nil
```

Comment:
There is a race condition here. Without any mutex there is a possibility of adding two producers.
Maybe you can create a singleton with sync.Once and with mutex delete/add producer when it is enabled/disabled. 
You can also create all entries once and simply based on boolean (memory/cpu) read those metrics.